pkg/parser/expression.go

```diff
@@ -0,0 +1,402 @@
+	return fc
+}
+
+func (p *Parser) parseExpr(scope *scope, prec precedence) Node {
+	var left Node
+	switch p.cur.Type {
+	case lexer.IDENT:
+		left = p.lookupVar(scope)
+	case lexer.STRING_LIT, lexer.NUM_LIT, lexer.TRUE, lexer.FALSE, lexer.LBRACKET, lexer.LCURLY:
+		left = p.parseLiteral(scope)
+	case lexer.BANG, lexer.MINUS:
+		left = p.parseUnaryExpr(scope)
+	case lexer.LPAREN:
+		left = p.parseGroupedExpr(scope)
+	default:
+		p.appendError("unexpected " + p.cur.FormatDetails())
+	}
+	for left != nil && !p.isAtExprEnd() && prec < p.curPrecedence() {
+		tt := p.cur.Type
+		switch {
+		case isBinaryOp(tt):
+			left = p.parseBinaryExpr(scope, left)
+		case tt == lexer.LBRACKET:
+			left = p.parserIndexExpr(scope, left)
+		case tt == lexer.DOT:
+			left = p.parserDotExpr(left)
+		default:
+			return left
+		}
+	}
+	return left // nil for previous error
+}
+
+func (p *Parser) isAtExprEnd() bool {
+	return p.isAtEOL() // TODO: in WS mode include WS.
+}
+
+func (p *Parser) parseUnaryExpr(scope *scope) Node {
+	tok := p.cur
+	unaryExp := &UnaryExpression{Token: tok, Op: op(tok)}
+	p.advance() // advance past operator
+	unaryExp.Right = p.parseExpr(scope, PREFIX)
+	if unaryExp.Right == nil {
+		return nil // previous error
+	}
+	p.validateUnaryType(unaryExp)
+	return unaryExp
+}
+
+func (p *Parser) parseBinaryExpr(scope *scope, left Node) Node {
+	tok := p.cur
+	expType := left.Type()
+	if isComparisonOp(tok.Type) {
+		expType = BOOL_TYPE
+	}
+	binaryExp := &BinaryExpression{Token: tok, T: expType, Op: op(tok), Left: left}
+	prec := p.curPrecedence()
+	p.advance() // advance past operator
+	binaryExp.Right = p.parseExpr(scope, prec)
+	if binaryExp.Right == nil {
+		return nil // previous error
+	}
+	p.validateBinaryType(binaryExp)
+	return binaryExp
+}
+
+func (p *Parser) parseGroupedExpr(scope *scope) Node {
+	p.advance() // advance past (
+	exp := p.parseTopLevelExpr(scope)
+	if !p.assertToken(lexer.RPAREN) {
+		return nil
+	}
+	p.advance() // advance past )
+	return exp
+}
+
+func (p *Parser) parserIndexExpr(scope *scope, left Node) Node {
+	tok := p.cur // TODO: ensure not prefixed by WS
+	p.advance()  // advance past [
+	leftType := left.Type().Name
+	if leftType != ARRAY && leftType != MAP && leftType != STRING {
+		p.appendErrorForToken("only array, string and map type can be indexed, found "+left.Type().Format(), tok)
+		return nil
+	}
+	index := p.parseTopLevelExpr(scope)
+	if !p.assertToken(lexer.RBRACKET) {
+		return nil
+	}
+	p.advance() // advance past ]
+	if leftType == ARRAY || leftType == STRING {
+		if index.Type() != NUM_TYPE {
+			p.appendError(leftType.String() + " index expects num, found " + index.Type().Format())
+			return nil
+		}
+		return &IndexExpression{Token: tok, Left: left, Index: index, T: left.Type().Sub}
+	}
+	if index.Type() != STRING_TYPE {
+		p.appendError("map index expects string, found " + index.Type().Format())
+		return nil
+	}
+	return &IndexExpression{Token: tok, Left: left, Index: index}
```

Comment:
should this be setting `T` like you do for the array/string indexed expression? If so, I'd reorder the above code to:
```
if (leftType == ARRAY || leftType == STRING) && index.Type() != NUM_TYPE {
    return error...
}
if leftType == MAP && index.Type() != STRING_TYPE {
    return error...
}
return &IndexExpression{Token: tok, Left: left, Index: index, T: left.Type().Sub}
```

Happy path and all, but I also find that error checking simpler to read - it spells out exactly what is required in one line each (i.e. array and string need a numeric index, map needs a string index)

Reply:
ok, that's an outright bug - sorry and thanks for finding. I need to do more type checking but was lazily assuming the evaluation stage will shake all of this out...
fixed